### ISSUE TITLE
Provide custom platform for D8 site created with Composer

### ISSUE THREAD

**Issue by startinggravity** [omega8cc/boa#1116]
I created a local instance of a D8 site using Composer to install it. Composer files and Drupal configuration files are in the project root, with the Drupal site being in /web. As I got ready to install in my BOA installation I discovered that could be a problem. 

I found in [#1065](https://github.com/omega8cc/boa/issues/1061) this statement:
> Custom builds/platforms created with drupal-composer don't work at the moment.

Is that still the case? Is there no way around me having to restructure my files? I hate to think so, but I'm at a loss for finding an alternative.

**Comment by Lowell20** [comment]
The only BOA method I found to work is the git clone drupal 8.x, for example, then modify the root composer.json to meet my needs. Don't forget to chmod -R 775 before verifying the platform.

**Comment by omega8cc** [comment]
It is a known issue, as explained at the top of: https://learn.omega8.cc/platforms

You could use some downloadable and supported D8 distros as an example of supported D8 directory structure: https://www.drupal.org/project/thunder

Note that this limitation should be removed in the near future: https://www.drupal.org/node/2736801

**Comment by dnotes** [comment]
Now that https://www.drupal.org/node/2838489 has been merged into Aegir 3.x, what else needs to be done before this is supported?

**Comment by omega8cc** [comment]
@dnotes We are testing this to make sure things are stable in the BOA specific environment, so it should be included in the next release.

**Comment by yaazkal** [comment]
It will be great, since it will allow [Commerce 2.x](https://github.com/drupalcommerce/commerce) to be used (any many more). I can see many projects going for this kind of implementation based on the [Composer template](https://github.com/drupal-composer/drupal-project) for Drupal projects.
It will be awesome if this can be included on a next release.

**Comment by yaazkal** [comment]
Hi I had time to test today and actually I had no issue installing Commerce 2.x (actually a private custom project based on [commerce/project-base](https://github.com/drupalcommerce/project-base)) on BOA 3.2.0 . I was able to clone the repository in the ~/static folder and then run `composer install --no-dev` on the cloned project. Just pointed the platform to the /web folder when I was creating the platform.

Only advice is to set the corresponding chmod 775 (at least for web/modules and web/sites) and also moving any profile from web/profiles/contrib outsite the contrib folder. Note that a tweak to the composer.json file can do the last thing automatically (I mean by the time of downloading the packages with `composer install`). So forking any [drupal-composer](https://github.com/drupal-composer/drupal-project) based project and doing the change may work (just for simplicity).

Don't forget to also use the drupal-scaffold plugin if you are not commiting those files (index.php update.php etc) to your composer project repository in order to get those files into your project by the time of running `composer install` and to allow BOA to recognize the platform. More info [here](https://github.com/drupal-composer/drupal-project#should-i-commit-the-scaffolding-files)

I had to test what happens with migrations, but at least I can breath for now since I'm working now on Drupal 8 projects that I need to host.

**Comment by omega8cc** [comment]
Thanks for sharing your experiment results @yaazkal -- we are working on it, and there is a progress also in the Aegir upstream.

**Comment by yaazkal** [comment]
@omega8cc thanks so much. By now, is there any strategy you are using or do you suggest for the sync folder? because generally speaking there is a `/sites/*/files/` inside .gitignore and BOA is generating the settings.php file pointing config to `sites/example.com/private/config/sync` wich will not be include as the `files` folder is not tracked on git, do you simply override the path on `local.settings.php` file?

**Comment by yaazkal** [comment]
@omega8cc thanks. I guess that the /bin and /drush folder for the generally used [drupal-project](https://github.com/drupal-composer/drupal-project) will not be necesary in a BOA environment, but how about the sync directory? I like the idea for it to be outside the /web and I can see that BOA is just using the private directory to store that config/sync folder, wich I think there is some nginx config for that folder to deny any access.

Just for curiosity, how does will be in future? you will support /web folder as mentioned by @dnotes or we just have to reformat our composer.json file?

**Comment by vebester** [comment]
I have managed to create the D8 platform pointing to /docroot with composer and lightning profile. It works, but /admin, /node/add pages are always redirected to <front>, /user pages are good

**Comment by yaazkal** [comment]
> It works, but /admin, /node/add pages are always redirected to front

Yes, for now the only way to use D8 in BOA is to have all drupal related files on root directory and no in subdirectory as _web_ or _docroot_. If you are using composer you should have to update the installation paths on the _composer.json_ file.
This is more an Aegir related issue and I guess there are some advances on the Aegir project side. We have to wait for a release and then for BOA to integrate it.

**Comment by vebester** [comment]
Yes, it works if all drupal stuff in a platform directory an no subdirectory as web or docroot is used. They work very slowly.

**Comment by ar-jan** [comment]
I can't get this to work, despite having the Drupal files in the platform's root dir:

```
cd ~/static
composer create-project drupal/drupal:8.4.0 drupal-8.4.0.1-dev
chmod -R 775 ~/static/drupal-8.4.0.1-dev
```

The drupal/drupal project uses repo root as docroot.

The platform can be added, but site install fails with:

> Drush command terminated abnormally due to an unrecoverable error.
> Error: Declaration of Drush\Command\DrushInputAdapter::hasParameterOption() must be compatible with Symfony\Component\Console\Input\InputInterface::hasParameterOption($values, $onlyParams = false) in /data/disk/o1/tools/drush/lib/Drush/Command/DrushInputAdapter.php, line 27

This looks like the problem caused by the Symfony update in 8.4 (https://www.drupal.org/node/2874827), but the solution is supposed to be to use Drush 8.1.15.

The Drush version on my BOA 3.2.1 install is 8.1.15-dev, which should be OK? Any clues?

**Comment by ar-jan** [comment]
Have any of you @yaazkal @vebester @omega8cc managed to get it to work with 8.4? Despite having Drupal in the root dir site-install fails for me with a Drush error (as above).

**Comment by omega8cc** [comment]
@ar-jan We have tested it with the VarBase distribution (and vanilla 8.4.0) yesterday, and it works with zero issues (but with BOA head only!) after building the stable VarBase platform with Composer, so the vendor dir is outside of docroot. Example: http://varbase.o1.quad.tt.o8.io

Make sure that you are using the platform/docroot as a path.

You will need the two patches linked below to make it work on BOA stable:

https://github.com/omega8cc/boa/commit/1dfa9e19ddb86c999f66f8a12ba8af5b73c146ec
https://github.com/omega8cc/hosting/commit/7eb8039fa33f1027616001c2618ccd2f9ab68135

**Comment by omega8cc** [comment]
We have built it with standard command:

`o1@quad:~/static$ composer create-project Vardot/varbase-project:^8.4.10 varbase-8.4.10 --no-dev --no-interaction`

**Comment by cwaone** [comment]
Hi everyone,

i have the same problem as @ar-jan - i updatet BOA yesterday to 3.2.1 - everything works ok - today i tried to add a new static platform for Drupal 8.4.4 - this worked without errors and the verify of the platform was also ok. But when i try to install a new site it breaks with the error mentioned above.

i checked my drush version, the symfony version, i added the two patches from @omega8cc from november - no chance to get this working - i also tried other Drupal 8.4 Version without any success. 

i also tried to install the platforms with composer or manually via sftp - no success - always the same error. do i miss something else? Drupal 8.3.7 is working perfect on that server.

thank you

**Comment by omega8cc** [comment]
@cwaone -- You would need to either wait for [BOA-3.2.2](https://github.com/omega8cc/boa/milestone/25), or apply the patches [we have linked above](https://github.com/omega8cc/boa/issues/1116#issuecomment-341143754) or just upgrade to BOA HEAD. Sorry for any possible confusion.

**Comment by yaazkal** [comment]
@ar-jan @cwaone I'm running 8.4.3 sites on BOA 3.2.1
Actually never faced the error you describe.

My workflow is to generate the project on my machine and just run `composer install --nodev` on the server (aftet git pull on the server). I have to point that I also manage the project using git and using the configuration commands to export and import configuration.

I can't remember if those sites where installed on 8.3.x on the server and later updated, which could be a workaround for you. Try installing an 8.3.x platform and site, and then change the composer file to update to 8.4.x and see if that works (`composer install --no-dev` then `drush @site.alias updb` and then `drush @site.alias cr`).

I have updated sites just by pulling the git repo and updating the platform code and also by creating a new platform and then migrating the site. Both works.

**Comment by yaazkal** [comment]
Just in case any one wants to use them, I've forked and edited two base projects to make Drupal 8 work on BOA:

Drupal 8: https://github.com/yaazkal/BOA-drupal-project
Drupal 8 + Commerce: https://github.com/yaazkal/BOA-commerce-project

**Comment by luxio** [comment]
@yaazkal thanks for sharing.

**Comment by yaazkal** [comment]
@luxio it's a pleasure!

**Comment by yaazkal** [comment]
Drupal Lightning: https://github.com/yaazkal/BOA-lightning-project

**Comment by cwaone** [comment]
Hi, first of all i can told you about my question few weeks ago, that after the update to 3.2.2 i could install drupal 8.4.4 without any problems. Installing, cloning, migrating - all works great :-)

Now, i have tried to install a new distro - varbase - i get the platform to verify correctly - but if i want to install a new site - with the included profile - the task breaks with error:

"Error: Allowed memory size of 536870912 bytes exhausted (tried to allocate 535103 bytes) in /data/disk/o1/static/varbase/docroot/core/lib/Drupal/Core/Database/Statement.php, line 59"

It´s just with this profile, a default vanilla installation works without any issues.

@omega8cc i used your command mentioned here https://github.com/omega8cc/boa/issues/1116#issuecomment-341144589 - did you try to install a varbase site in your platform? did i miss something in my server config?

**Comment by omega8cc** [comment]
@cwaone What is the RAM size on your system?

**Comment by cwaone** [comment]
@omega8cc on my test system first i have 4GB - after your question i upgraded the vm to 8 GB - the next installation also failed with the same error mentioned above.

**Comment by omega8cc** [comment]
@fotiDim Yes, we have disabled this by default (although it can be enabled with Drush variable easily). Please check https://github.com/omega8cc/boa/blob/master/docs/COMPOSER.txt

**Comment by omega8cc** [comment]
@startinggravity -- BOA already supports /web or /docroot style structure. Please check: https://github.com/omega8cc/boa/blob/master/docs/COMPOSER.txt

**Comment by 7o7o** [comment]
Adding to @cwaone comments, Varbase platforms can be succesfully created and also sites can be installed on the platform and everything works correctly. 

Where I had problems after creating the project with composer was when trying to add any other module to the platform with "composer require". I get the "PHP Fatal error:  Allowed memory size of 2147483648 bytes exhausted (tried to allocate 4096 bytes) in phar:///usr/local/bin/composer/src/Composer/DependencyResolver/Solver.php on line 223". Even with 8GB of RAM size.

But after trying, i managed it to work by copying the composer.json and doing a "composer install" on a new project folder, with the modules i wanted to add to the Varbase project. So, if anybody else is having problems doing composer require on a Varbase profile, try doing composer install with the modified version of composer.json. Platform was succesfully added to the Aegir panel also after that.

**Comment by omega8cc** [comment]
@7o7o We don't recommend managing codebase with composer on any production system like it was drush, because it's not. It's not even about the crazy memory issues often caused my PHP memory leaks with Composer, but it's about development habits, like with previous-era makefiles -- you shouldn't try to modify already deployed platform, only build a new one from updated makefile or composer.json etc.

**Comment by 7o7o** [comment]
Oh yeah @omega8cc, i meant to create a new platform with every change. Sorry I didnt expressed myself very good, english is not my mother tongue. I always create the platform with composer first and then i add it to Aegir. Thanks for the advice, you are great.

**Comment by ar-jan** [comment]
I saw on Twitter that you can run Drupal 9 on BOA. I'm trying this with the officially recommended composer template, but while it works with v8.9, it fails with v9.0:

    composer create-project drupal/recommended-project:8.9.x ~/static/platforms/drupal-8.9.1-dev --no-dev --no-interaction
    chmod -R 775 ~/static/platforms/drupal-8.9.1-dev

Then adding the platform in Aegir UI with path `platforms/drupal-8.9.1-dev/web` works.

But the same with `composer create-project drupal/recommended-project ~/static/platforms/drupal-9.0.1-dev --no-dev --no-interaction` and platform path in Aegir `platforms/drupal-9.0.1-dev/web` fails with these errors:

> Unable to include file packages9inc or packagesinc from !path.
> Unable to include the drupal engine packages from /data/disk/o1/.drush/sys/provision/platform/drupal.
> Drush command terminated abnormally due to an unrecoverable error.
> Error: Uncaught Error: Call to undefined function _provision_drupal_search_paths() in /data/disk/o1/.drush/sys/provision/platform/provision_drupal.drush.inc:713
> Stack trace:
> #0 /data/disk/o1/.drush/sys/provision/platform/provision_drupal.drush.inc(685): _provision_drupal_find_modules('base', '')
> #1 /data/disk/o1/.drush/sys/provision/platform/provision_drupal.drush.inc(646): _provision_find_packages('base')
> #2 /data/disk/o1/.drush/sys/provision/platform/verify.provision.inc(160): provision_find_packages()
> #3 /data/disk/o1/tools/drush/includes/command.inc(422): drush_provision_drupal_pre_provision_verify()
> #4 /data/disk/o1/tools/drush/includes/command.inc(231): _drush_invoke_hooks(Array, Array)
> #5 /data/disk/o1/tools/drush/includes/command.inc(199): drush_command()
> #6 /data/disk/o1/tools/drush/lib/Drush/Boot/BaseBoot.php(67): drush_dispatch(Array)
> #7 /data/disk/o1/tools/drush/includes/preflight.inc(67): Drush\Boot\BaseBoot->bootstrap_and_ in /data/disk/o1/.drush/sys/provision/platform/provision_drupal.drush.inc, line 713
> The external command could not be executed due to an application error.

Any extra tricks needed for D9?

**Comment by omega8cc** [comment]
@ar-jan There are new requirements: you must run PHP 7.3 for CLI and FPM, you must upgrade to Percona 5.7 (or MariaDB 10.3 or 10.4, although we no longer recommend MariaDB because of its politics since 10.3 breaking paths for libraries, so fast sql dump doesn’t work etc). Once you have barracuda running Percona 5.7 and PHP 7.3, it’s time to upgrade also octopus with `octopus up-head o1 force`. Now you can install Drupal 9, although we have tested only the tarball from d.o without Composer which requires 4GB RAM at minimum for CLI alone.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by ar-jan** [comment]
Just to make sure, how to upgrade to Percona? Is it a matter of configuring
    _DB_SERIES=5.7
    _DB_SERVER=Percona
And then running upgrade? Is `barracuda up-head system` sufficient, or should it be a full upgrade?

**Comment by omega8cc** [comment]
@ar-jan Yes, that's it:

`sed -i "s/^_DB_SERIES=.*/_DB_SERIES=5.7/g"  /root/.barracuda.cnf`
`sed -i "s/^_DB_SERVER=.*/_DB_SERVER=Percona/g"  /root/.barracuda.cnf`
`barracuda up-head system`

**Comment by omega8cc** [comment]
It's now possible to run just `barracuda up-head system percona` or even `barracuda up-head percona`
